### PR TITLE
Operations needed for TPCC.

### PR DESCRIPTION
**Pull request by amlatyrngom**
Brings in Insert, Update & Delete from TPL.

### CODE REVIEW

File: src/include/execution/sql/updater.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+   * @return Whether update was successful.
+   */
+  bool TableUpdate(storage::TupleSlot table_tuple_slot);
+
+  /**
+   * Delete a tuple from the table.
+   * @param table_tuple_slot tuple slot of the tuple.
+   * @return Whether delete was successful.
+   */
+  bool TableDelete(storage::TupleSlot table_tuple_slot);
```
Comment (jrolli) [inline src/include/execution/sql/updater.h]: Does the `Updater` need `Delete` and `Insert` functions?  I think the binder and/or optimizer should handle rewriting an update on an indexed attribute to being a `select`, `delete`, `insert` sequence.  If that is the case, then these functions are unnecessary.

File: src/execution/sql/inserter.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  auto txn = exec_ctx_->GetTxn();
+  table_redo_ = txn->StageWrite(exec_ctx_->DBOid(), table_oid_, pri);
+  table_pr_ = table_redo_->Delta();
+  return table_pr_;
+}
+
+storage::ProjectedRow *Inserter::GetIndexPR(catalog::index_oid_t index_oid) {
+  // cache?
+  auto index = GetIndex(index_oid);
+  auto index_pri = index->GetProjectedRowInitializer();
```
Comment (jrolli) [inline src/execution/sql/inserter.cpp]: Not sure why the compiler didn't catch this, but this is an unused variable and should be removed.

File: src/execution/sql/deleter.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  for (auto index_oid : index_oids) {
+    max_pr_size_ = std::max(max_pr_size_, GetIndex(index_oid)->GetProjectedRowInitializer().ProjectedRowSize());
+  }
+  index_pr_buffer_ = exec_ctx->GetMemoryPool()->AllocateAligned(max_pr_size_, sizeof(uint64_t), true);
+}
+
+Deleter::~Deleter() { exec_ctx_->GetMemoryPool()->Deallocate(index_pr_buffer_, max_pr_size_); }
+
+common::ManagedPointer<storage::index::Index> execution::sql::Deleter::GetIndex(catalog::index_oid_t index_oid) {
+  auto iter = index_cache_.find(index_oid);
```
Comment (mbutrovich) [inline src/execution/sql/deleter.cpp]: I don't think we should be caching catalog lookups at this level. We should cache at the Catalog level.

File: src/include/execution/sql/deleter.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+  catalog::table_oid_t table_oid_;
+  exec::ExecutionContext *exec_ctx_;
+  common::ManagedPointer<terrier::storage::SqlTable> table_;
+
+  void *index_pr_buffer_;
+  uint32_t max_pr_size_;
+  storage::ProjectedRow *index_pr_{nullptr};
+
+  std::map<terrier::catalog::index_oid_t, common::ManagedPointer<storage::index::Index>> index_cache_;
```
[the remark this answers is not in the document]
Comment (amlatyrngom) [reply src/include/execution/sql/deleter.h]: Actually, this cache is an overkill. We only need one `Index*` member variable to avoid multiple lookups. Will be fixed.

File: src/execution/sql/updater.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+storage::ProjectedRow *Updater::GetTablePRForColumns(const std::vector<catalog::col_oid_t> &col_oids) {
+  TERRIER_ASSERT(!col_oids.empty(), "Should have at least one column oid");
+  storage::ProjectedRowInitializer pri = table_->InitializerForProjectedRow(col_oids);
+  auto txn = exec_ctx_->GetTxn();
+  table_redo_ = txn->StageWrite(exec_ctx_->DBOid(), table_oid_, pri);
+  return table_redo_->Delta();
+}
+
+storage::ProjectedRow *Updater::GetIndexPR(catalog::index_oid_t index_oid) {
+  TERRIER_ASSERT(is_index_key_update_, "Should not be called as index key is not being updated");
```
[the remark this answers is not in the document]
Comment (amlatyrngom) [reply src/execution/sql/updater.cpp]: It's definitely not strong enough in the general case. It relies on optimizer/codegen handling this correctly. Checking at runtime requires traversing expression trees and checking updated columns.

File: src/include/execution/sql/inserter.h [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  storage::TupleSlot table_tuple_slot_;
+  storage::RedoRecord *table_redo_{nullptr};
+
+  storage::ProjectedRow *table_pr_{nullptr};
+
+  void *index_pr_buffer_;
+  uint32_t max_pr_size_;
+  storage::ProjectedRow *index_pr_{nullptr};
+
+  std::map<terrier::catalog::index_oid_t, common::ManagedPointer<storage::index::Index>> index_cache_;
```
[the remark this answers is not in the document]
Comment (amlatyrngom) [reply src/include/execution/sql/inserter.h]: See other answer.

File: src/include/execution/sql/projected_row_wrapper.h [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+   * @return The typed value at the current iterator position in the column
+   */
+  template <typename T, bool Nullable>
+  const T *Get(uint16_t col_idx, bool *null) const {
+    // NOLINTNEXTLINE: bugprone-suspicious-semicolon: seems like a false positive because of constexpr
+    if constexpr (Nullable) {
+      TERRIER_ASSERT(null != nullptr, "Missing output variable for NULL indicator");
+      *null = pr_->IsNull(col_idx);
+    }
+    return reinterpret_cast<T *>(pr_->AccessWithNullCheck(col_idx));
```
[the remark this answers is not in the document]
Comment (amlatyrngom) [reply src/include/execution/sql/projected_row_wrapper.h]: Yes. There should be an `AccessWithoutNullCheck` call to speed up the case where we know it's not nullable.
Will fix it.

File: src/include/execution/sql/updater.h [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  uint32_t CalculateMaxIndexPRSize();
+  void *GetMaxSizedIndexPRBuffer();
+
+  void PopulateAllColOids(std::vector<catalog::col_oid_t> *all_col_oids);
+
+  storage::ProjectedRow *GetTablePRForColumns(const std::vector<catalog::col_oid_t> &col_oids);
+
+  catalog::table_oid_t table_oid_;
+  std::vector<catalog::col_oid_t> col_oids_;
+  std::vector<catalog::col_oid_t> all_col_oids_;
```
[the remark this answers is not in the document]
Comment (amlatyrngom) [reply src/include/execution/sql/updater.h]: Not needed if we can rely on the optimizer to pass in all column oids.

File: src/execution/sql/inserter.cpp [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  index_pr_buffer_ = exec_ctx->GetMemoryPool()->AllocateAligned(max_pr_size_, sizeof(uint64_t), true);
+}
+
+Inserter::~Inserter() { exec_ctx_->GetMemoryPool()->Deallocate(index_pr_buffer_, max_pr_size_); }
+
+storage::ProjectedRow *Inserter::GetTablePR() {
+  // We need all the columns
+  storage::ProjectedRowInitializer pri = table_->InitializerForProjectedRow(col_oids_);
```
Comment (mbutrovich) [inline src/execution/sql/inserter.cpp]: Can we stash this in the Inserter? This is very expensive to create for every PR.

File: src/execution/vm/bytecode_generator.cpp [hunk 10]
```diff
@@ -1449,148 +1467,310 @@ void BytecodeGenerator::VisitBuiltinIndexIteratorCall(ast::CallExpr *call, ast::
       Emitter()->Emit(Bytecode::IndexIteratorFree, iterator);
       break;
     }
-    case ast::Builtin::IndexIteratorGetTinyInt: {
-      LocalVar val = ExecutionResult()->GetOrCreateDestination(ast::BuiltinType::Get(ctx, ast::BuiltinType::Integer));
-      auto col_idx = static_cast<uint16_t>(call->Arguments()[1]->As<ast::LitExpr>()->Int64Val());
-      Emitter()->EmitIndexIteratorGet(Bytecode::IndexIteratorGetTinyInt, val, iterator, col_idx);
+    case ast::Builtin::IndexIteratorGetPR: {
+      ast::Type *pr_type = ast::BuiltinType::Get(ctx, ast::BuiltinType::ProjectedRow);
+      LocalVar pr = ExecutionResult()->GetOrCreateDestination(pr_type);
```
Comment (pmenon) [inline src/execution/vm/bytecode_generator.cpp]: The return type of the call should have been set during semantic analysis. So you can simplify to `ExecutionResult()->GetOrCreateDestination(call->Type())`

File: src/execution/vm/bytecode_generator.cpp [hunk 11]
```diff
@@ -1449,148 +1467,310 @@ void BytecodeGenerator::VisitBuiltinIndexIteratorCall(ast::CallExpr *call, ast::
...
+      break;
+    }
+    case ast::Builtin::InserterGetIndexPRBind: {
+      ast::Type *pr_type = ast::BuiltinType::Get(ctx, ast::BuiltinType::ProjectedRow);
+      LocalVar pr = ExecutionResult()->GetOrCreateDestination(pr_type);
+      ast::Identifier index_name = call->Arguments()[1]->As<ast::LitExpr>()->RawStringVal();
+      auto ns_oid = exec_ctx_->GetAccessor()->GetDefaultNamespace();
+      auto index_oid = exec_ctx_->GetAccessor()->GetIndexOid(ns_oid, index_name.Data());
+
+      Emitter()->EmitInserterGetIndexPR(Bytecode::InserterGetIndexPR, pr, inserter, !index_oid);
```
Comment (pmenon) [inline src/execution/vm/bytecode_generator.cpp]: Why `!index_oid` ?

File: src/include/execution/vm/bytecode_handlers.h [hunk 12]
```diff
@@ -1321,215 +1325,232 @@ VM_OP void OpIndexIteratorPerformInit(terrier::execution::sql::IndexIterator *it
...
+    *out = terrier::execution::sql::StringVal(reinterpret_cast<const char *>(varlen->Content()), varlen->Size());
+  }
 }
 
-VM_OP_HOT void OpIndexIteratorGetBigIntNull(terrier::execution::sql::Integer *out,
-                                            terrier::execution::sql::IndexIterator *iter, uint16_t col_idx) {
-  // Read
-  bool null = false;
-  auto *ptr = iter->Get<int64_t, true>(col_idx, &null);
+// Inserter Calls
```
Comment (pmenon) [inline src/include/execution/vm/bytecode_handlers.h]: Try to keep comment style consistent within the file.

File: src/execution/ast/ast.cpp [hunk 13]
```diff
@@ -49,6 +49,13 @@ bool Expr::IsIntegerLiteral() const {
   return false;
 }
 
+bool Expr::IsBooleanLiteral() const {
```
Comment (pmenon) [inline src/execution/ast/ast.cpp]: Is this used in this PR? If not, remove it.

File: sample_tpl/updater.tpl [hunk 14]
```diff
@@ -0,0 +1,116 @@
+
+struct output_struct {
```
Comment (pmenon) [inline sample_tpl/updater.tpl]: Add high-level comments indicating its purpose. The final count at the bottom is unclear.

File: src/execution/vm/vm.cpp [hunk 15]
```diff
@@ -1401,56 +1427,268 @@ void VM::Interpret(const uint8_t *ip, Frame *frame) {
...
+  }
+
+  OP(IndexIteratorGetSlot) : {
+    auto *slot = frame->LocalAt<storage::TupleSlot *>(READ_LOCAL_ID());
+    auto *iter = frame->LocalAt<sql::IndexIterator *>(READ_LOCAL_ID());
+    OpIndexIteratorGetSlot(slot, iter);
+    DISPATCH_NEXT();
+  }
+
+  /////////////////////////////////
```
Comment (pmenon) [inline src/execution/vm/vm.cpp]: Try to be consistent with the comment style in the rest of the file.

File: test/execution/inserter_test.cpp [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+
+// NOLINTNEXTLINE
+TEST_F(InserterTest, SimpleInserterTest) {
+  //
+  auto table_oid = exec_ctx_->GetAccessor()->GetTableOid(NSOid(), "test_1");
+  Inserter inserter(exec_ctx_.get(), table_oid);
+  auto table_pr = inserter.GetTablePR();
+  auto schema = exec_ctx_->GetAccessor()->GetSchema(table_oid);
+
+  *reinterpret_cast<int16_t *>(table_pr->AccessForceNotNull(3)) = 15;
```
Comment (yeshengm) [inline test/execution/inserter_test.cpp]: This doesn't seem to match the sema of test_1?

File: src/execution/vm/bytecode_generator.cpp (hunk 10 above)
Comment (pmenon) [reply src/execution/vm/bytecode_generator.cpp]: Still valid.

File: src/execution/vm/bytecode_generator.cpp [hunk 17]
```diff
@@ -1449,148 +1467,239 @@ void BytecodeGenerator::VisitBuiltinIndexIteratorCall(ast::CallExpr *call, ast::
...
+      UNREACHABLE("Impossible bytecode");
+    }
+  }
+}
+
+void BytecodeGenerator::VisitBuiltinPRCall(ast::CallExpr *call, ast::Builtin builtin) {
+  // First argument is always a projected row
+  LocalVar pr = VisitExpressionForRValue(call->Arguments()[0]);
+  ast::Context *ctx = call->GetType()->GetContext();
+  // TODO(Amadou): I am using VisitExpressionForLValue because it seems to involve one less copy.
```
Comment (pmenon) [inline src/execution/vm/bytecode_generator.cpp]: Remove comment.

File: src/execution/vm/vm.cpp (hunk 15 above)
Comment (pmenon) [reply src/execution/vm/vm.cpp]: Still valid.

File: src/include/execution/sql/projected_row_wrapper.h [hunk 18]
```diff
@@ -0,0 +1,65 @@
+#pragma once
+#include "execution/util/execution_common.h"
+#include "storage/projected_row.h"
+
+namespace terrier::execution::sql {
+
+/**
+ * Wrapper around projected rows.
```
Comment (pmenon) [inline src/include/execution/sql/projected_row_wrapper.h]: Let's merge this into `ProjectedRow` unless there's a specific reason for it.

File: src/include/execution/sql/projected_row_wrapper.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (pmenon) [reply src/include/execution/sql/projected_row_wrapper.h]: Is @jrolli's comment still valid?

File: src/include/execution/util/hash.h [hunk 19]
```diff
@@ -23,7 +23,7 @@ enum class HashMethod : uint8_t { Fnv1, Crc, Murmur2, xxHash3 };
  * for arithmetic values (integers and floats), and generic versions for longer
  * buffers (strings, c-strings, and opaque buffers).
  */
-class Hasher {
+class EXPORT Hasher {
```
Comment (pmenon) [inline src/include/execution/util/hash.h]: Why is this needed in this PR?

File: src/include/execution/vm/bytecode_emitter.h [hunk 20]
```diff
@@ -430,38 +421,34 @@ class BytecodeEmitter {
...
-   * @param bytecode bytecode used to free an index iterator
-   * @param iter index iterator to free
+   * Initialize a StringVal from a char array
+   * @param bytecode bytecode to emit
+   * @param out where to store the result
+   * @param length length of the string
+   * @param data pointer to the char array
    */
-  void EmitIndexIteratorFree(Bytecode bytecode, LocalVar iter);
+  void EmitInitString(Bytecode bytecode, LocalVar out, uint64_t length, uintptr_t data);
```
Comment (pmenon) [inline src/include/execution/vm/bytecode_emitter.h]: This will have to be removed ...

File: src/include/execution/vm/bytecode_handlers.h [hunk 21]
```diff
@@ -1321,215 +1323,202 @@ VM_OP void OpIndexIteratorPerformInit(terrier::execution::sql::IndexIterator *it
 
 VM_OP_HOT void OpIndexIteratorScanKey(terrier::execution::sql::IndexIterator *iter) { iter->ScanKey(); }
 
-VM_OP_HOT void OpIndexIteratorAdvance(bool *has_more, terrier::execution::sql::IndexIterator *iter) {
-  *has_more = iter->Advance();
-}
+VM_OP_HOT void OpIndexIteratorScanAscending(terrier::execution::sql::IndexIterator *iter) { iter->ScanAscending(); }
```
Comment (pmenon) [inline src/include/execution/vm/bytecode_handlers.h]: Are these really hot? They're implementations are not inlined, implying that they're not hot. Either (1) move `Scan*()` implementations to header or (2) mark these are WARM or move into `cpp`.

I'd suggest (2) for now and inline as we observe performance issues.

File: src/include/execution/vm/bytecode_handlers.h [hunk 22]
```diff
@@ -1321,215 +1323,202 @@ VM_OP void OpIndexIteratorPerformInit(terrier::execution::sql::IndexIterator *it
...
-  // Read
-  bool null = false;
-  auto *ptr = iter->Get<int8_t, true>(col_idx, &null);
-
-  // Set
-  out->is_null_ = null;
-  out->val_ = null ? 0 : *ptr;
+VM_OP_HOT void OpPRSetVarlenNull(terrier::execution::sql::ProjectedRowWrapper *pr, uint16_t col_idx,
+                                 terrier::execution::sql::StringVal *val) {
+  if (val->len_ <= terrier::storage::VarlenEntry::InlineThreshold()) {
```
Comment (pmenon) [inline src/include/execution/vm/bytecode_handlers.h]: You need to move this logic out of the handlers into `StringVal`. It doesn't belong there either, but it's better than here.

File: util/execution/table_generator/table_generator.cpp [hunk 23]
```diff
@@ -14,14 +14,13 @@
 
 namespace terrier::execution::sql {
 template <typename T>
-T *TableGenerator::CreateNumberColumnData(Dist dist, uint32_t num_vals, uint64_t min, uint64_t max) {
-  static uint64_t serial_counter = 0;
+T *TableGenerator::CreateNumberColumnData(ColumnInsertMeta *col_meta, uint32_t num_vals) {
```
Comment (pmenon) [inline util/execution/table_generator/table_generator.cpp]: I don't see any functional change here. Any other reason to include in this PR?

File: util/include/execution/table_generator/table_generator.h [hunk 24]
```diff
@@ -73,6 +73,10 @@ class TableGenerator {
      * Distribution of values
      */
     Dist dist_;
+    /**
```
Comment (pmenon) [inline util/include/execution/table_generator/table_generator.h]: Ditto for this file.

File: sample_tpl/updater-2.tpl [hunk 25]
```diff
@@ -0,0 +1,198 @@
+fun index_count_2(execCtx: *ExecutionContext, key : int64, key2 : int64) -> int64 {
```
Comment (pmenon) [inline sample_tpl/updater-2.tpl]: Comment this file and the complex logic pieces within.

File: src/include/execution/vm/bytecode_handlers.h [hunk 26]
```diff
@@ -1321,215 +1323,202 @@ VM_OP void OpIndexIteratorPerformInit(terrier::execution::sql::IndexIterator *it
...
 
-VM_OP_HOT void OpIndexIteratorGetBigIntNull(terrier::execution::sql::Integer *out,
-                                            terrier::execution::sql::IndexIterator *iter, uint16_t col_idx) {
+VM_OP_HOT void OpPRGetDateNull(terrier::execution::sql::Date *out, terrier::execution::sql::ProjectedRowWrapper *pr,
+                               uint16_t col_idx) {
   // Read
   bool null = false;
-  auto *ptr = iter->Get<int64_t, true>(col_idx, &null);
+  auto *ptr = pr->Get<uint32_t, true>(col_idx, &null);
+  TERRIER_ASSERT(ptr != nullptr, "Null pointer when trying to read date");
```
Comment (mbutrovich) [inline src/include/execution/vm/bytecode_handlers.h]: You can't do this assert if the type is nullable (assuming that's what the function signature means). Null attributes return nullptr when reading from the storage layer.

File: util/execution/table_generator/table_generator.cpp (hunk 23 above)
Comment (mbutrovich) [reply util/execution/table_generator/table_generator.cpp]: Per Amadou: needed to get rid of the `static` to be able to reuse it multiple times in test cases.

File: src/include/execution/util/hash.h (hunk 19 above)
Comment (mbutrovich) [reply src/include/execution/util/hash.h]: Solved a compilation issue on macOS, I think.

File: src/include/execution/vm/bytecode_emitter.h (hunk 20 above)
Comment (mbutrovich) [reply src/include/execution/vm/bytecode_emitter.h]: This will be removed in a future PR that @amlatyrngom is working on to implement parameters.

File: src/include/execution/sql/projected_row_wrapper.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (jrolli) [reply src/include/execution/sql/projected_row_wrapper.h]: The concern is still valid, but it is not a correctness issue.  We can leave a TODO for now, but this will likely need to be fixed in the long run since this will probably be on the critical path of every OLTP query.
